File: dotnet/Shared/PasswordGenerator.cs
namespace PasswordGenerator.Shared;

/// <summary>
/// Pure business logic for generating passwords.
/// This class has no external dependencies and focuses solely on password generation.
/// </summary>
public sealed class PasswordGenerator
{
    private const string LowercaseChars = "abcdefghijklmnopqrstuvwxyz";
    private const string UppercaseChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private const string NumericChars = "0123456789";
    private const string SpecialChars = "!@#$%^&*()-_=+[]{}|;:,.<>?";

    private readonly IRandomSource _randomSource;

    /// <summary>
    /// Initializes a new instance of the PasswordGenerator class.
    /// </summary>
    /// <param name="randomSource">Source of randomness for password generation.</param>
    /// <exception cref="ArgumentNullException">Thrown when randomSource is null.</exception>
    public PasswordGenerator(IRandomSource randomSource)
    {
        _randomSource = randomSource ?? throw new ArgumentNullException(nameof(randomSource));
    }

    /// <summary>
    /// Generates a password according to the specified configuration.
    /// </summary>
    /// <param name="config">Password configuration.</param>
    /// <returns>The generated password.</returns>
    /// <exception cref="ArgumentNullException">Thrown when config is null.</exception>
    public string Generate(PasswordConfig config)
    {
        ArgumentNullException.ThrowIfNull(config);

        var charPool = BuildCharacterPool(config.UseSpecialChars);
        var password = new char[config.Length];

        // Generate random indices
        var randomBytes = _randomSource.GetRandomBytes(config.Length);

        for (int i = 0; i < config.Length; i++)
        {
            var index = randomBytes[i] % charPool.Length;
            password[i] = charPool[index];
        }

        return new string(password);
    }

    /// <summary>
    /// Calculates the strength of a password.
    /// </summary>
    /// <param name="password">The password to evaluate.</param>
    /// <returns>The password strength level.</returns>
    public static PasswordStrength CalculateStrength(string password)
    {
        if (string.IsNullOrEmpty(password))
            return PasswordStrength.Weak;

        bool hasLower = password.Any(c => char.IsLower(c));
        bool hasUpper = password.Any(c => char.IsUpper(c));
        bool hasDigit = password.Any(c => char.IsDigit(c));
        bool hasSpecial = password.Any(c => SpecialChars.Contains(c));

        int strength = 0;
        if (hasLower) strength++;
        if (hasUpper) strength++;
        if (hasDigit) strength++;
        if (hasSpecial) strength++;

        if (password.Length < 8)
            return PasswordStrength.Weak;

        if (strength <= 2)
            return PasswordStrength.Weak;

        if (strength == 3)
            return PasswordStrength.Medium;

        return PasswordStrength.Strong;
    }

    /// <summary>
    /// Builds the character pool based on configuration.
    /// </summary>
    private static string BuildCharacterPool(bool useSpecialChars)
    {
        var pool = LowercaseChars + UppercaseChars + NumericChars;
        if (useSpecialChars)
            pool += SpecialChars;
        return pool;
    }
}

File: python/shared/password_generator.py
"""Pure business logic for password generation."""

from .models import PasswordConfig, PasswordStrength
from .interfaces import IRandomSource

LOWERCASE_CHARS = "abcdefghijklmnopqrstuvwxyz"
UPPERCASE_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
NUMERIC_CHARS = "0123456789"
SPECIAL_CHARS = "!@#$%^&*()-_=+[]{}|;:,.<>?"


class PasswordGenerator:
    """
    Pure business logic for generating passwords.
    This class has no external dependencies and focuses solely on password generation.
    """

    def __init__(self, random_source: IRandomSource) -> None:
        """
        Initialize the password generator.

        Args:
            random_source: Source of randomness for password generation.

        Raises:
            ValueError: When random_source is None.
        """
        if random_source is None:
            raise ValueError("random_source cannot be None")
        self._random_source = random_source

    def generate(self, config: PasswordConfig) -> str:
        """
        Generates a password according to the specified configuration.

        Args:
            config: Password configuration.

        Returns:
            The generated password.

        Raises:
            ValueError: When config is None.
        """
        if config is None:
            raise ValueError("config cannot be None")

        char_pool = self._build_character_pool(config.use_special_chars)
        random_bytes = self._random_source.get_random_bytes(config.length)

        password = ""
        for i in range(config.length):
            index = random_bytes[i] % len(char_pool)
            password += char_pool[index]

        return password

    @staticmethod
    def calculate_strength(password: str) -> PasswordStrength:
        """
        Calculates the strength of a password.

        Args:
            password: The password to evaluate.

        Returns:
            The password strength level.
        """
        if not password:
            return PasswordStrength.WEAK

        has_lower = any(c.islower() for c in password)
        has_upper = any(c.isupper() for c in password)
        has_digit = any(c.isdigit() for c in password)
        has_special = any(c in SPECIAL_CHARS for c in password)

        strength = sum([has_lower, has_upper, has_digit, has_special])

        if len(password) < 8:
            return PasswordStrength.WEAK
        if strength <= 1:
            return PasswordStrength.WEAK

        if strength == 4:
            return PasswordStrength.STRONG

        return PasswordStrength.MEDIUM

    @staticmethod
    def _build_character_pool(use_special_chars: bool) -> str:
        """Build the character pool based on configuration."""
        pool = LOWERCASE_CHARS + UPPERCASE_CHARS + NUMERIC_CHARS
        if use_special_chars:
            pool += SPECIAL_CHARS
        return pool
